### PR TITLE
feat: context-aware gated N-gram embedding with causal dilated conv

### PR DESCRIPTION
**Pull request by Ngk03**
## Summary

Replace simple averaging and sigmoid/softmax gates with **DeepSeek V4-style sqrt(softplus) scoring + competitive normalization** for N-gram Embedding fusion. Uses causal depthwise dilated convolutions to compute context-aware gate weights, enabling dynamic token-adaptive specialization across unigram and ngram sub-table experts.

## Architecture

```
word_emb [B, S, H]
    │
    ├── expert_0: word_emb (unigram)
    ├── expert_1..4: bigram sub-table projections
    └── expert_5..8: trigram sub-table projections
    │
    └── Causal Depthwise Dilated Conv × N levels
         │  dilation=1 (unigram context)  → Linear → 1 score
         │  dilation=2 (bigram context)   → Linear → K scores
         │  dilation=3 (trigram context)  → Linear → K scores
         ▼
    sqrt(softplus(logits)) → normalize → gate_weights [B, S, 9]
         │
         ▼ output = Σ(gate_i × expert_i)
```

## Key Design

- **sqrt(softplus)** scoring: `sqrt(log(1 + exp(x)))` — non-negative, non-saturating, validated by DeepSeek V4
- **Competitive normalization**: L1-normalize creates zero-sum competition among experts
- **Multi-scale context**: Different dilation per ngram level (d=1,2,3) captures context at appropriate scales
- **Causal**: Left-padding + trim ensures future tokens cannot influence current position
- **Near-uniform init**: Gate projection ~ Normal(0, 0.01) → initial gate ≈ 1/9 per expert, compatible with pretrained checkpoints

## Changes

| File | Change |
|------|--------|
| `ngram_gated_embedding.py` | **New** — Core `NgramGatedEmbedding` module |
| `ngram_embedding.py` | **New** — LongCat-style N-gram embedding + `forward_per_table()` |
| `language_model_embedding.py` | Integrate ngram + gated path |
| `transformer_config.py` | Add 9 config fields (ngram + gate) |

## Config

```yaml
ngram_embedding_enabled: true
ngram_vocab_size_ratio: 12.2
ngram_emb_neighbor_num: 3
ngram_emb_split_num: 4
ngram_gate_enabled: true
ngram_gate_conv_kernel_size: 7
ngram_gate_route_scale: 1.0
```

## Overhead

- Params: ~123K (negligible vs model)
- FLOPs: ~1B/batch (negligible vs transformer ~1T/layer)
- Memory: bounded by [B, S, 9, H] expert stack (temporary)

## Related

- Based on N-gram Embedding infrastructure from PR #956
- Replaces sigmoid gate (v1) which collapsed to near-constant behavior
- Inspired by DeepSeek V4 MoE gating (arxiv 2606.19348)

### CODE REVIEW

File: src/paddlefleet/models/common/embeddings/ngram_embedding.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        return vocab_mods
+
+    def _compute_shifted_ids(self, input_ids: Tensor) -> dict:
+        """Compute right-shifted input_ids for each offset k."""
+        shifted_ids = {}
+        B, S = input_ids.shape
+
+        for k in range(1, self.n):
+            shifted = paddle.full([B, k], self.pad_token_id, dtype=input_ids.dtype)
+            shifted = paddle.concat([shifted, input_ids[:, :S - k]], axis=1)
```
Comment (liuhao2638) [inline src/paddlefleet/models/common/embeddings/ngram_embedding.py:112]: P1 **优先级：P1**

这里在 `k >= S` 时会生成比输入更长的 shifted ids。默认 `ngram_emb_neighbor_num=3` 会计算 `k=2`，如果推理/增量预填充传入单 token 序列，`paddle.full([B, 2], ...)` 与 `input_ids[:, :S-k]` 拼接后宽度为 2，后续 `_compute_ngram_hash_ids()` 会把它和宽度为 1 的 `input_ids` 相加，N-gram 路径直接 shape mismatch。请保证每个 shifted tensor 始终是 `[B, S]`。

```suggestion
            if k < S:
                shifted = paddle.full([B, k], self.pad_token_id, dtype=input_ids.dtype)
                shifted = paddle.concat([shifted, input_ids[:, :S - k]], axis=1)
            else:
                shifted = paddle.full([B, S], self.pad_token_id, dtype=input_ids.dtype)
```

File: src/paddlefleet/models/common/embeddings/ngram_embedding.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        else:
+            self.emb_dim = self.hidden_size // num_embedders
+
+        self.normalizer = 1 + num_embedders
+
+        embedders = []
+        post_projs = []
+        for i in range(num_embedders):
+            emb_vocab_size = int(self.m + i * 2 + 1)
+            emb = nn.Embedding(emb_vocab_size, self.emb_dim)
```
Comment (liuhao2638) [inline src/paddlefleet/models/common/embeddings/ngram_embedding.py:73]: P1 **优先级：P1**

默认 `ngram_vocab_size_ratio=12.2` 时，这里每个 sub-table 都会创建约 `12.2 * vocab_size` 行；默认 `split=4, neighbor=3` 一共 8 张表，并且自动 `emb_dim = hidden_size // 8`。以常见 `vocab_size=100000, hidden_size=4096` 计算，仅这些 N-gram embedding 权重就是约 `8 * 1,220,000 * 512 = 4,997,152,768` 个参数，bf16 也要约 9.3 GiB，还没有任何 TP 分片。这和 PR 描述的 “Params: ~123K” 差距很大，启用默认配置会让 embedding 层不可用。请改成有界/可分片的参数规模，或者把默认 ratio/维度设计调整到和目标开销一致，并补充构造期参数量测试。

可按现有并行 embedding 的方式拆分每张 N-gram 表，例如：

```python
# 伪代码：每张表按 tensor parallel vocab 维分片，而不是在每个 rank 复制完整表
emb = tensor_parallel.VocabParallelEmbedding(
    num_embeddings=emb_vocab_size,
    embedding_dim=self.emb_dim,
    init_method=config.embedding_init_method,
    reduce_scatter_embeddings=False,
    config=config,
)
```

**Review by liuhao2638 — CHANGES_REQUESTED** [review]
已完成本轮审查，发现需要先修复的阻塞问题，具体见行级评论。

- P3 **优先级：P3** 非行级：PR 描述不属于 diff 行。当前 PR body 没有按仓库模板补充 `PR Category` / `PR Types` / `是否引起精度变化`，也缺少实际测试计划；请补齐这些信息，并同步修正参数开销说明，避免后续评估风险时缺少依据。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
